recipes_source/recipes/tuning_guide.py

```diff
@@ -0,0 +1,376 @@
+# `Tensor Cores <https://www.nvidia.com/en-us/data-center/tensor-cores/>`_.
+#
+# Support for `channels_last` is new and experimental, but it's expected to work
+# for standard computer vision models (e.g. ResNet-50, SSD). To convert models
+# to `channels_last` format follow
+# `Channels Last Memory Format Tutorial <https://pytorch.org/tutorials/intermediate/memory_format_tutorial.html>`_.
+# The tutorial includes a section on
+# `converting existing models <https://pytorch.org/tutorials/intermediate/memory_format_tutorial.html#converting-existing-models>`_.
+
+###############################################################################
+# Parallelize kernel execution
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# When short kernels are executed in serial, the execution efficiency becomes
+# low due to time needed to schedule and launch work on the accelerator.
+# `JIT <https://pytorch.org/docs/stable/jit.html>`_ can help alleviate this cost
+# but its benefit is restricted when inputs are not feedforward between kernels.
+#
+# When kernels do not have input dependency, aggregating the kernels with the
+# same input dimensions and executing work in parallel helps to reduce the
+# number of kernels to execute and increase the resource occupancy within each
+# kernel execution. Good examples to apply kernel vectorization are parameter
+# optimizers and modules with multiple heads (e.g. self-attention).
+
+# Build list of outputs
+out = []
+for block in layer_blocks:
+    out.append(block(x))
+
+# Construct stacked tensor
+out_ = torch.stack(out, dim=0)
+
+# Execute functions that are shared by all blocks in parallel
+out_ = element_wise_layers(out_)
+
+# Re-distribute outputs back to each head
+out = torch.split(out_, dim=0)
+
+# Continue processing layer sequences of each layer
+
+###############################################################################
+# Checkpoint intermediate buffers
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# Buffer checkpointing is a technique to mitigate the memory capacity burden of
+# model training. Instead of storing inputs of all layers to compute upstream
+# gradients in back-prop, it stores the inputs of a few layers and the others
+# are recomputed during backward pass. The reduced memory requirements enables
+# increasing the batch size that can improve utilization.
+#
+# Checkpointing targets should be selected carefully. The best is not to store
+# large layer outputs that have small re-computation cost. The example target
+# layers are activation functions (e.g. ``ReLU``, ``Sigmoid``, ``Tanh``),
+# up/down sampling and matrix-vector operations with small accumulation depth.
+#
+# PyTorch supports a native
+# `torch.utils.checkpoint <https://pytorch.org/docs/stable/checkpoint.html>`_
+# API to automatically perform checkpointing and recomputation.
+
+###############################################################################
+# Pre-allocate memory in case of variable input length
```

Comment:
This is a GPU-specific optimization because CPU is unlikely to run out of memory, and does not rely on caching allocator.

Reply:
I'll move it to GPU-specific section